p2p/net/conn/dial.go

```diff
@@ -265,3 +209,100 @@ func MultiaddrNetMatch(tgt ma.Multiaddr, srcs []ma.Multiaddr) ma.Multiaddr {
 	}
 	return nil
 }
+
+type ProtoDialer interface {
```

Comment:
This might belong somewhere better than here though, and the interface probably could use some tweaks, but this is the composition that i'm wanting.